adapters/smaato/smaato.go

```diff
@@ -49,240 +49,374 @@ type siteExtData struct {
 	reqJSON, err := json.Marshal(request)
 	if err != nil {
-		errs = append(errs, err)
-		return nil, errs
+		return nil, err
 	}
 
-	uri := a.URI
-
 	headers := http.Header{}
 	headers.Add("Content-Type", "application/json;charset=utf-8")
 	headers.Add("Accept", "application/json")
 
-	return []*adapters.RequestData{{
+	return &adapters.RequestData{
 		Method:  "POST",
-		Uri:     uri,
+		Uri:     adapter.endpoint,
 		Body:    reqJSON,
 		Headers: headers,
-	}}, errs
+	}, nil
 }
 
-// MakeBids unpacks the server's response into Bids.
-func (a *SmaatoAdapter) MakeBids(internalRequest *openrtb2.BidRequest, externalRequest *adapters.RequestData, response *adapters.ResponseData) (*adapters.BidderResponse, []error) {
-	if response.StatusCode == http.StatusNoContent {
-		return nil, nil
-	}
-
-	if response.StatusCode == http.StatusBadRequest {
-		return nil, []error{&errortypes.BadInput{
-			Message: fmt.Sprintf("Unexpected status code: %d. Run with request.debug = 1 for more info", response.StatusCode),
-		}}
-	}
-
-	if response.StatusCode != http.StatusOK {
-		return nil, []error{fmt.Errorf("unexpected status code: %d. Run with request.debug = 1 for more info", response.StatusCode)}
-	}
-
-	var bidResp openrtb2.BidResponse
-	if err := json.Unmarshal(response.Body, &bidResp); err != nil {
-		return nil, []error{err}
+func getAdMarkupType(response *adapters.ResponseData, adMarkup string) (adMarkupType, error) {
+	if admType := adMarkupType(response.Headers.Get("X-Smt-Adtype")); admType != "" {
+		return admType, nil
+	} else if strings.HasPrefix(adMarkup, `{"image":`) {
+		return smtAdTypeImg, nil
+	} else if strings.HasPrefix(adMarkup, `{"richmedia":`) {
+		return smtAdTypeRichmedia, nil
+	} else if strings.HasPrefix(adMarkup, `<?xml`) {
+		return smtAdTypeVideo, nil
+	} else {
+		return "", &errortypes.BadServerResponse{
+			Message: fmt.Sprintf("Invalid ad markup %s.", adMarkup),
+		}
 	}
+}
 
-	bidResponse := adapters.NewBidderResponseWithBidsCapacity(5)
-
-	for _, sb := range bidResp.SeatBid {
-		for i := 0; i < len(sb.Bid); i++ {
-			bid := sb.Bid[i]
-
-			markupType, markupTypeErr := getAdMarkupType(response, bid.AdM)
-			if markupTypeErr != nil {
-				return nil, []error{markupTypeErr}
-			}
-
-			var markupError error
-			bid.AdM, markupError = renderAdMarkup(markupType, bid.AdM)
-			if markupError != nil {
-				return nil, []error{markupError}
-			}
-
-			bidType, bidTypeErr := markupTypeToBidType(markupType)
-			if bidTypeErr != nil {
-				return nil, []error{bidTypeErr}
-			}
+func (adapter *SmaatoAdapter) getTTLFromHeaderOrDefault(response *adapters.ResponseData) int64 {
+	ttl := int64(300)
 
-			bidResponse.Bids = append(bidResponse.Bids, &adapters.TypedBid{
-				Bid:     &bid,
-				BidType: bidType,
-			})
+	if expiresAtMillis, err := strconv.ParseInt(response.Headers.Get("X-Smt-Expires"), 10, 64); err == nil {
+		nowMillis := adapter.clock.Now().UnixNano() / 1000000
+		ttl = (expiresAtMillis - nowMillis) / 1000
+		if ttl < 0 {
```

Comment:
Can you please add a JSON test for this condition?

Reply:
sure, added